File: puzzle8-1.py
import sys

path = sys.stdin.readline().rstrip()
sys.stdin.readline()
nodes = {}
for line in sys.stdin:
    node, lr = line.rstrip().split(' = ')
    l, r = lr.strip('()').split(', ')
    nodes[node] = {'L':l, 'R':r}
    # print(node, l, r)

step = 0
cur = 'AAA'
pathlen = len(path)
while cur != 'ZZZ':
    cur = nodes[cur][path[step % pathlen]]
    step += 1
print(step)

File: lua/puzzle8-1.lua
require "utils"

local path = io.read()
io.read()
print("path", path)

local nodes = {}
for line in io.lines() do
	local name, targ = unpack(string.split(line, " = "))
	assert(name and targ)
	targ = string.strip(targ, "()")
	local l, r = unpack(string.split(targ, ", "))
	assert(l and r)
	nodes[name] = {L=l, R=r}
end

local cur = "AAA"
local len = #path
local i = 0
while cur ~= "ZZZ" do
	local j = (i % len) + 1
	local dir = string.sub(path, j, j)
	cur = nodes[cur][dir]
	i = i + 1
end
print(i)
